### ISSUE TITLE
Make Unified Search configurable

### ISSUE THREAD

**Issue by andreas-p** [nextcloud/server#24056]
As covered in #23835, searching for files may take a very long time, making file search quite unusable. But it also has the side effect, that searching for something else will create collateral damage by overwhelming the database with unnecessary, very expensive queries, which stack up with every new keystroke. While the tests in #23835 where performed with my test user, who has 4 shares to a very big storage, other users have an average of 25 shares, with 55 shares max. If that user types 5 keys in the search box, he will trigger 4x55 queries running 1.2s each...

As a quick fix, I removed SearchProvider registration for the files app. 

Each search provider should be configurable (at least enable/disable systemwide)

**Comment by ruedigerkupper** [comment]
I must say that a rarely (next to never) need  or want "unified search". If a search for a contact, I go to the contacts app (and don't want to see files or events). If I search for an event in my calendar, I naturally open the calendar app and search for it (and don't want to see files, contacts, tasks or whatever).
In general, I always know *what* I'm searching for. I never open Nextcloud and say, "let's search for 'abc' and see what I get! Oh, there's a file. Oh, and there's a contact, too! How exciting!" This just does not happen. In may opinion, having search results displayed all over Nextcloud is a disadvantage, it makes searching much harder and much slower. It is a step back in productivity.
So, yes, it would be nice if I could just disable global search and go with local search as I expect searching to work.
I can imagine that global search could be a nice feature when searching from the "dashboard" app. This app is designed to gather results from all over Nextcloud. But "files" sould search for files, "contacts" for contacts, "calendar" for events, "bookmarks" for bookmarks etc.

**Comment by ruedigerkupper** [comment]
After all, the aim of the search feature is to give the user what he's searching for, as fast as possible. So, if the user already knows *what* she's searching for, it is unreasonable to ignore this fact. It may be a good option being able to broaden the search later, if the searched item is not found. But starting with the broadest search possible is mostly not helpful.

**Comment by szaimen** [comment]
cc @nextcloud/designers is this feasible and does it make sense?

**Comment by andreas-p** [comment]
Actually, the the proposal was issued trying to make unified search usable again without dropping it rightaway, while I completely agree with @ruedigerkupper : I can't recall any incident that I ever wanted to use unified search in my life, neither on NC nor on Windows or MacOS. So I could live with dedicated search tools only quite fine as well. If they exist.

**Comment by jospoortvliet** [comment]
So isn't this already implemented? You can pick what you want to search for, and even use an easy keyword:
![ezgif-6-84f14fbbedfc](https://user-images.githubusercontent.com/551757/125066797-b2e43480-e0b3-11eb-858f-41951cf58601.gif)

I would love to see a full-screen search and all that, but if you want to search in files only - just type `in:files searchterm` and you get what you want...

**Comment by andreas-p** [comment]
See #23835 why I need to disable filesearch completely. I can't let this up to the user.

**Comment by ruedigerkupper** [comment]
> So isn't this already implemented? You can pick what you want to search for, and even use an easy keyword:
> ![ezgif-6-84f14fbbedfc](https://user-images.githubusercontent.com/551757/125066797-b2e43480-e0b3-11eb-858f-41951cf58601.gif)
> 
> I would love to see a full-screen search and all that, but if you want to search in files only - just type `in:files searchterm` and you get what you want...

Right, so in the files app "in:files" should be the default, because I have already decided I am looking for files and stated rhat by going to the files app. Unspecific search may be a nice idea for the dashboard, but not dedicated apps.

**Comment by szaimen** [comment]
> Right, so in the files app "in:files" should be the default, because I have already decided I am looking for files and stated rhat by going to the files app. Unspecific search may be a nice idea for the dashboard, but not dedicated apps.

Sounds like a good proposal to me 👍

**Comment by andreas-p** [comment]
@szaimen any solution that doesn't allow the administrator to disable file search is not a solution to my problem. As described in the first post, file search from unified search is fundamentally broken, so I can't allow users to use it right now. 

Your change of the title is thus not reflecting the original purpose of this thread.

**Comment by szaimen** [comment]
Lets make clear that the search performance has already improved by a lot in the last releases and we won't allow disabling search!

But you are right that I am misusing the issue for a different solution that you've intended.

**Comment by andreas-p** [comment]
The file search still performs a full table scan, which isn't going to work on big databases. I can allow searches in the current directory only (which already may contain some thousand files/directories).

**Comment by andreas-p** [comment]
22.1 doesn't multiply queries by shares any more, which is a good thing. Still, it's unified search, when only local search is needed.

**Comment by marcelklehr** [comment]
This is what most services do nowadays:

![image](https://user-images.githubusercontent.com/986878/130356047-2cfb1453-4c55-4d8f-8cb4-e76cc205657a.png)

**Comment by R0Wi** [comment]
@jospoortvliet in my case i'm using the [fulltextsearch](https://github.com/nextcloud/fulltextsearch) app which provides a much quicker way to search for files and integrates in unified search, too. Of course you also need a lot more infrastructure but i think the admins who also installed this app would be happy to disable at least the files searchprovider. So if it's a desireable feature to disable/enable searchproviders globally i'd volunteer to create a pull requests for this 👍

**Comment by szaimen** [comment]
Hey @juliushaertl and @ChristophWurst I'd like to tackle this. Could you please give me some code pointers where this needs to be implemented? Thanks! :)

**Comment by Spartachetto** [comment]
@szaimen take into account that possibly you'd better contact @Aeredren because he could be working on the same thing. For reference #29614 and https://help.nextcloud.com/t/looking-forward-to-develop-sorts-files-visualisation-improvement-and-filters/128314 .

**Comment by szaimen** [comment]
@Aeredren how far did you get with this until now? Is this issue (searchnng in the current app by default) something that you've already implemented in the planned enhancements?

**Comment by Aeredren** [comment]
Hi @szaimen , I did not tackle the issue of "unified search" vs "search in current app", because I was focusing on search modal (filters) for files. (which does not use search provider but OCP\Files\Search) Once upon a time I was planning to look into it but my time runs short so feel free to work on it :)

**Comment by kyteinsky** [comment]
@szaimen Is it up for grabs?

**Comment by szaimen** [comment]
> @szaimen Is it up for grabs?

yes it is. Feel free to work on it :)

**Comment by kyteinsky** [comment]
dropping this since it would be worked on here: #39162

**Comment by pjrobertson** [comment]
The new unified search in NC 28 (see https://github.com/nextcloud/server/issues/39162) is a HUGE improvement! However, it still searches ALL apps by default as opposed to the current app. I agree that searching current app would still be the preferred option. 

The reddit example in the [comment above](https://github.com/nextcloud/server/issues/24056#issuecomment-903264928) is a good example, as is GitHub:

<img width="335" alt="Screen Region 2023-12-22 at 17 56 43" src="https://github.com/nextcloud/server/assets/150431/d356e5b0-e4fe-43fd-94dd-f7e1cfd20eca">


For the new unified search, I think it would look something like what was originally posted by marcoambrosini here: https://github.com/nextcloud/server/issues/39162#issuecomment-1635223315

**Comment by szaimen** [comment]
There was a PR by @marcoambrosini which would be a big improvement IMO :)
https://github.com/nextcloud/server/pull/41609
